java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java

```diff
@@ -0,0 +1,176 @@
+      completeSubstitution(currentSubstitution, formalType, argType);
+    } else if (formalType.isArray()) {
+      JavaType newArgType = null;
+      if (argType.isArray()) {
+        newArgType = ((ArrayJavaType) argType).elementType;
+      } else if (variableArity) {
+        newArgType = leastUpperBound(remainingArgTypes);
+      }
+      if (newArgType != null) {
+        JavaType formalElementType = ((ArrayJavaType) formalType).elementType;
+        TypeSubstitution newSubstitution = inferTypeSubstitution(method, currentSubstitution, formalElementType, newArgType, variableArity, remainingArgTypes);
+        return mergeTypeSubstitutions(currentSubstitution, newSubstitution);
+      }
+    } else if (formalType.isParameterized()) {
+      List<JavaType> formalTypeSubstitutedTypes = ((ParametrizedTypeJavaType) formalType).typeSubstitution.substitutedTypes();
+      if (argType.isParameterized()) {
+        List<JavaType> argTypeSubstitutedTypes = ((ParametrizedTypeJavaType) argType).typeSubstitution.substitutedTypes();
+        TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, argTypeSubstitutedTypes);
+        return mergeTypeSubstitutions(currentSubstitution, newSubstitution);
+      } else if (isRawTypeOfType(argType, formalType) || isNullType(argType)) {
+        List<JavaType> fakeTypes = new ArrayList<>(formalTypeSubstitutedTypes.size());
+        for (int j = 0; j < formalTypeSubstitutedTypes.size(); j++) {
+          fakeTypes.add(symbols.objectType);
+        }
+        TypeSubstitution newSubstitution = inferTypeSubstitution(method, formalTypeSubstitutedTypes, fakeTypes);
+        return mergeTypeSubstitutions(currentSubstitution, newSubstitution);
+      } else if (argType.isSubtypeOf(formalType.erasure()) && argType.isClass()) {
+        for (JavaType superType : ((ClassJavaType) argType).symbol.superTypes()) {
+          if (sameErasure(formalType, superType)) {
+            return inferTypeSubstitution(method, currentSubstitution, formalType, superType, variableArity, remainingArgTypes);
+          }
+        }
+      }
+    } else if (formalType.isTagged(JavaType.WILDCARD)) {
+      JavaType inferedFromArg = argType;
+      if (argType.isTagged(JavaType.WILDCARD)) {
+        inferedFromArg = ((WildCardType) argType).bound;
+      }
+      TypeSubstitution newSubstitution = inferTypeSubstitution(method, currentSubstitution, ((WildCardType) formalType).bound, inferedFromArg, variableArity, remainingArgTypes);
+      return mergeTypeSubstitutions(currentSubstitution, newSubstitution);
+    } else {
+      // nothing to infer for simple class types or primitive types
+    }
+    return currentSubstitution;
+  }
+
+  private static boolean sameErasure(JavaType formalType, JavaType superType) {
+    return formalType.erasure() == superType.erasure();
+  }
+
+  private boolean isNullType(JavaType argType) {
+    return argType == symbols.nullType;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)